packages/next/src/server/app-render/app-render.tsx

```diff
@@ -336,90 +336,157 @@ function createFlightDataResolver(ctx: AppRenderContext) {
-      getDynamicParamFromSegment: getDynamicParamFromSegment,
-      appUsingSizeAdjustment: appUsingSizeAdjustment,
-    })
+  const { seedData, styles } = await createComponentTree({
+    ctx,
+    createSegmentPath: (child) => child,
+    loaderTree: tree,
+    parentParams: {},
+    firstItem: true,
+    injectedCSS,
+    injectedJS,
+    injectedFontPreloadTags,
+    rootLayoutIncluded: false,
+    asNotFound: asNotFound,
+    metadataOutlet: <MetadataOutlet />,
+  })
 
-    const { seedData, styles } = await createComponentTree({
-      ctx,
-      createSegmentPath: (child) => child,
-      loaderTree: loaderTreeToRender,
-      parentParams: {},
-      firstItem: true,
-      injectedCSS,
-      injectedJS,
-      injectedFontPreloadTags,
-      rootLayoutIncluded: false,
-      asNotFound: props.asNotFound,
-      metadataOutlet: <MetadataOutlet />,
-    })
+  return (
+    <>
+      {styles}
+      <AppRouter
+        buildId={ctx.renderOpts.buildId}
+        assetPrefix={ctx.assetPrefix}
+        initialCanonicalUrl={urlPathname}
+        // This is the router state tree.
+        initialTree={initialTree}
+        // This is the tree of React nodes that are seeded into the cache
+        initialSeedData={seedData}
+        initialHead={
+          <>
+            {ctx.res.statusCode > 400 && (
+              <meta name="robots" content="noindex" />
+            )}
+            {/* Adding requestId as react key to make metadata remount for each render */}
+            <MetadataTree key={ctx.requestId} />
+          </>
+        }
+        globalErrorComponent={GlobalError}
+      />
+    </>
+  )
+}
 
-    return (
-      <>
-        {styles}
-        <AppRouter
-          buildId={ctx.renderOpts.buildId}
-          assetPrefix={ctx.assetPrefix}
-          initialCanonicalUrl={urlPathname}
-          // This is the router state tree.
-          initialTree={initialTree}
-          // This is the tree of React nodes that are seeded into the cache
-          initialSeedData={seedData}
-          initialHead={
-            <>
-              {ctx.res.statusCode > 400 && (
-                <meta name="robots" content="noindex" />
-              )}
-              {/* Adding requestId as react key to make metadata remount for each render */}
-              <MetadataTree key={ctx.requestId} />
-            </>
-          }
-          globalErrorComponent={GlobalError}
-        />
-      </>
-    )
-  }, options)
+type ReactServerErrorProps = {
+  tree: LoaderTree
+  ctx: AppRenderContext
+  preinitScripts: () => void
+  errorType: 'not-found' | 'redirect' | undefined
+}
+async function ReactServerError({
+  tree,
+  ctx,
+  preinitScripts,
+  errorType,
+}: ReactServerErrorProps) {
+  const {
+    getDynamicParamFromSegment,
+    query,
+    providedSearchParams,
+    appUsingSizeAdjustment,
+    componentMod: { AppRouter, GlobalError },
+    staticGenerationStore: { urlPathname },
+    requestId,
+    res,
+  } = ctx
+
```

Comment:
I've extracted the Render and Error RSC components into static functions accepting props rather than constructed functions consuming a closure. The actual contents on what is rendered should not have changed (apologies for how annoying this will be to check in review)